### ISSUE TITLE
Allow using MainContext from "glib" crate

### ISSUE THREAD

**Issue by mitchhentges** [jnqnfe/pulse-binding-rust#24]
This is a mirror of the [TODO in `pulse-sys-mainloop-glib`](https://github.com/jnqnfe/pulse-binding-rust/blob/caf0a340ef0f2f72492efdc58200f392aea1f656/pulse-sys-mainloop-glib/src/lib.rs#L25).

I'd like to be able to use `MainContext` from the `glib` crate as a parameter for [`Mainloop::new(...)` in `pulse-binding-glib`](https://docs.rs/libpulse-glib-binding/2.6.0/libpulse_glib_binding/struct.Mainloop.html#method.new).

I'd be interested in drawing up a PR for this, but I'm not sure how to peel the `GMainContext` out of the `glib` crate. `MainContext` is defined as:
```
pub struct MainContext(Shared<glib_sys::GMainContext>);
```
I'm not sure how to get at the `Shared` structure, since it's private. Once that's accomplished, though, I'm assuming I can just `shared.borrow_mut()` to get at the `GMainContext` inside?

**Comment by jnqnfe** [comment]
I'll take a look at the glib crate, if it's "the" goto crate for this then the right change will be to add it as a dependency and use its context type directly.

for the time being, all you need to do is directly transmute (an unsafe operation) a ref/pointer to one type as a ref/pointer to the other. (you know that my type is meant as an opaque alias).

**Comment by jnqnfe** [comment]
update: so after I posted that I looked into the glib crates a little and found that `glib` and `glib-sys` are both by far the most popular on crates.io and authored by the same people, so excellent candidates for mine to make use of.

I have now released new versions of my glib-mainloop crates which make use of `glib-sys::GMainContext`

However, we have a problem. Although you can use `glib-sys` with my crates, use of `glib` is an issue. We **need** to be able to get a raw `GMainContext` pointer from `glib` to feed through to the PulseAudio `pa_glib_mainloop_new` function, and I cannot see any available means for accessing this from `glib`. There's a ton of abstraction with macros and traits going on in `glib` which makes it a little hard to follow and to thus be certain its not possible but I failed to uncover anything obvious in the API docs. And no, we cannot just transmute a ref to `glib::MainContext` since it is a completely different object, same goes for the type it directly wraps.

I'm going to search through the `glib` bug listing and post my own if I don't find one for this already, to explain the situation and need for access to the raw pointer. It may involve waiting for them to ship a new release with such access provided, if they even agree to do so.

If they fail to give access to the pointer, there's nothing I can do on my side save for possibly re-implementing the entire PA glib-mainloop system library in Rust as a replacement, which is a no go.

**Comment by mitchhentges** [comment]
Thanks for digging further in! Considering that the underlying `pulseaudio` and `glib` software assume that the "main context" can be reused, I'm optimistic that the `glib` crate team will be aligned with us :smile:

**Comment by sdroege** [comment]
As discussed in the glib issue, `glib::translate::ToGlibPtr::to_glib_none()` is what you want to get a non-owned pointer of the main context. `main_context.to_glib_none().0` in your case.

If you need an owned pointer, `to_glib_full()` but make sure that something takes ownership of it and unrefs it afterwards.

**Comment by jnqnfe** [comment]
@sdroege, thanks, that works :)

okay, new version of the binding just released making use of `glib::MainContext`! :)

**Comment by mitchhentges** [comment]
Thanks all :smile:
